lightning/src/ln/channel.rs

```diff
@@ -1721,9 +1789,48 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 				removed_outbound_total_msat += htlc.amount_msat;
 			}
 		}
-		if htlc_inbound_value_msat + msg.amount_msat + self.value_to_self_msat > (self.channel_value_satoshis - Channel::<ChanSigner>::get_remote_channel_reserve_satoshis(self.channel_value_satoshis)) * 1000 + removed_outbound_total_msat {
-			return Err(ChannelError::Close("Remote HTLC add would put them under their reserve value"));
+
+		let pending_value_to_self_msat =
+			self.value_to_self_msat + htlc_inbound_value_msat - removed_outbound_total_msat;
+		let pending_remote_value_msat =
+			self.channel_value_satoshis * 1000 - pending_value_to_self_msat;
+		if pending_remote_value_msat < msg.amount_msat {
+			return Err(ChannelError::Close("Remote HTLC add would overdraw remaining funds"));
+		}
+
+		let remote_commit_tx_fee_msat = if self.channel_outbound { 0 } else {
+			// +1 for this HTLC.
+			self.next_remote_commit_tx_fee_msat(1)
+		};
+		if pending_remote_value_msat - msg.amount_msat < remote_commit_tx_fee_msat {
+			return Err(ChannelError::Close("Remote HTLC add would not leave enough to pay for fees"));
```

Comment:
This is the spec deviation right ? Please at least add a comment for it and modify commit message to say we fail channel. Anyway that may not conjugate well with anchor output, where once implemented a initiator-sender consider it's always good to send a HTLC if above reserve because you can always bump ?

Reply:
Sorry, don't believe this is a deviation. Curious why you think it's a deviation? 
From the BOLT 2:
```
update_add_htlc
A sending node:
    if it is responsible for paying the Bitcoin fee:
        MUST NOT offer amount_msat if, after adding that HTLC to its commitment transaction, it cannot pay the fee for either the local or remote commitment transaction at the current feerate_per_kw while maintaining its channel reserve
```

Reply:
I can modify the error message to spell out `"Remote HTLC add would not leave enough to pay for on-chain commitment transaction fees"`?

Reply:
AFAICT, this is a requirement on the _sender_ but here we're at HTLC reception. It makes sense as that way you avoid to close channel even if sender violates spec. I think channel can still be unstuck if this HTLC is failed and sender-initiator balance is increased again without waiting feerate decrease.

Reply:
Maybe misunderstood is due to the spec saying

```
update_add_htlc
A receiving node:
     receiving an amount_msat that the sending node cannot afford at the current `feerate_per_kw` (while maintaining its channel reserve):
            SHOULD fail the channel
```

It doesn't require to enforce fee buffer spikes.

Reply:
Ah but there's no fee spike buffer check here?

Here's the deviating fee spike buffer check: https://github.com/rust-bitcoin/rust-lightning/pull/577/files#diff-79ec46664bd0cf85b6be0ddb567527a7R1811-R1818

Reply:
> Maybe misunderstood is due to the spec saying
> 
> ```
> update_add_htlc
> A receiving node:
>      receiving an amount_msat that the sending node cannot afford at the current `feerate_per_kw` (while maintaining its channel reserve):
>             SHOULD fail the channel
> ```
> 
> It doesn't require to enforce fee buffer spikes.

Right, this check is intended to purely perform the check that you quote above here ^

Reply:
After IRC discussion, this is right, I confused myself with what this check was enforcing, but still you should have point out the _receiver_ spec-side :p

Reply:
That's fair, lol

Reply:
Saw the IRC discussion -- I'll add a clarifying comment